src/qibolab/instruments/qm/program/instructions.py

```diff
@@ -51,6 +61,12 @@ def play(args: ExecutionArguments):
     Should be used inside a ``program()`` context.
     """
     qua.align()
+
+    # keep track of ``Align`` command that were already played
+    # because the same ``Align`` will appear on multiple channels
+    # in the sequence
+    processed_aligns = set()
```

Comment:
> It's not that elegant from the point of view of the types (though it could be made more elegant - but it's not needed), still it should be much simpler (and you avoid iterating the whole sequence every time that you encounter an align, also because you know they are supposed to be consecutive, otherwise it would be a bug).

Indeed, my main issue with this is that it is not that elegant and may require more code from the driver side, like having a QM internal sequence, while I am trying to rely on qibolab primitives as much as possible. Also, the solution with the set shouldn't be much overhead in terms of performance and it is also more localized, just a few additional lines of code in a single place. The preprocess would probably have to happen somewhere else and copy the whole sequence, just to handle `Align` which is a very special and potentially rare case.

Reply:
The set is not the problem, but rather
https://github.com/qiboteam/qibolab/blob/433a2c2497ebb585166bc6f1806cea6d6e6bc941/src/qibolab/instruments/qm/program/instructions.py#L82
which requires iterating the whole sequence for every `Align` (for each ID, not each combination `(ch, Align)`, of course).

I was not proposing a separate sequence, just a single (inelegant) function doing:
```py
def process_aligns(sequence: PulseSequence) -> list[Union[tuple[ChannelId, PulseLike], tuple[Literal["align"], set[ChannelId]]]]:
    new = []
    align_channels = set()
    def align():
        if len(align_channels) > 0:
            new.append(("align", align_channels))
            align_channels.clear()
    for (ch, p) in sequence:
        if isinstance(p, Align):
            align_channels.add(ch)
        else:
            align()
            new.append((ch, p))
    align()
    return new
```
You can tell it's not elegant from the return type (though it could be written better than that), but it requires a single iteration, without any QM-specific object (it could even be part of the sequence API...).

However, as I said, it's negligible. I'm just obsessed...

Reply:
I see, thanks for the clarification. That is fine with me and I could add it to the code. Maybe a simplification would be to use the actual `Align` object in the tuple, instead of `Literal["align"]` and have the following return type:
```py
list[tuple[Union[ChannelId, set[ChannelId]], PulseLike]]
```
It isn't much simpler, it's just that the second tuple element is always `PulseLike`. We could even get rid of the `Union` if we convert the first element to be a `list` or `set` always, with single element for non-aligns.

My issue with all these solutions is that, although they are more efficient when `Align`s are present, they are slightly less efficient when `Align`s are not present. For example in this case we'd have to copy the sequence even when `Align` is not present. We could check and only call `process_aligns` if they are actually present, but even this check would be worst case O(n) (`n=len(sequence)`) since we'd have to check the whole sequence. In summary, the proposed solution would be O(2n) in all cases, while the current implementation is O((m + 1)n) (m=number of aligns), so if this is correct it depends on what we expect more often. For example the current compiler doesn't use `Align` (and most qibocal routines will not either) so if we optimize for circuits the current implementation may actually be better in most cases. Of course negligibly better, but since we are discussing this anyway...

Reply:
Yes, in practice the trade-off is even less clear, because the operations involved are not all equivalent: in one case you're just copying some `Align` (though iterating more `PulseLike`, if many `Align` are present), but the other is possibly doing fewer operations, but always copying the whole sequence. So, if copying were much more expensive than iterating, we had a winner...

In any case, as I said, it's just for the sake of discussion. The truth is that this part is not the performance bottleneck.
(already loading the whole platform, creating many objects, and parsing the whole `parameters.json`, irrespectively of the sequence, should be much more relevant - and still negligible, since most of the execution time in current experiments will be always the actual runtime on the device, and then the networking required)